src/main/scala/arcana/empress/Lexer.scala

```diff
@@ -0,0 +1,90 @@
+        case (locatedToken, next) => parseLine(parsed :+ locatedToken, next)
+      }
+    }
+
+  private def createLocation(start: Position, end: Position) =
+    if (options.enableLocationInfo) Some(Location(start, end, options.filename))
+    else None
+
+  private def parsePunctuation(punc: Char, position: Position) = 
+    (PunctuationToken(punc), createLocation(position, position :+ 1))
+
+  private def parseStringLiteral(str: String, start: Position) = {
+    val end = str.indexOf('"', start.column + 1) + 1
+    if (end == 0)
+      ((ErrorToken("missing \" in the string literal"),
+        createLocation(start, Position(start.line, str.length()))), str.length())
+    else
+      ((LiteralToken(str.substring(start.column, end)), createLocation(start, Position(start.line, end))), end)
+  }
+
+  private def parseNumericLiteral(str: String, start: Position, allowDotOrE: Boolean = true): (LocatedToken, Int) = {
+    val num = str.substring(start.column).takeWhile((c) => c.isDigit)
+    val end = start :+ num.length()
+
+    if ((str(end.column) == '.' || str(end.column) == 'e' || str(end.column) == 'E') &&
+         end.column + 1 != str.length() && str(end.column + 1).isDigit) {
+      parseNumericLiteral(str, Position(start.line, end.column + 1), false) match {
+        case ((ErrorToken(info), Some(loc)), next) => ((ErrorToken(info), createLocation(start, loc.end)), next)
+        case ((ErrorToken(info), None), next) => ((ErrorToken(info), None), next)
+        case ((LiteralToken(num2), Some(loc)), next) => ((LiteralToken(num + str(end.column) + num2), createLocation(start, loc.end)), next)
+        case ((LiteralToken(num2), None), next) => ((LiteralToken(num + str(end.column) + num2), None), next)
+        case _ => throw new AssertionError("fatal error during parsing numeric literal.") // this will not happen
+      }
+    }
+    else if (punctuations.contains(str(end.column)) && str(end.column) != '(' && str(end.column) != '[' && str(end.column) != '{')
+      ((LiteralToken(num), createLocation(start, end)), end.column)
+    else // e.g. 123abc
+      ((ErrorToken("wrong numeric expression"), createLocation(start, end)), end.column)
+  }
+
+  private def parseIdentifierAndKeyword(str: String, start: Position) = {
+    val res = str.substring(start.column).takeWhile((c) => 
+      !punctuations.contains(c))
+    val end = start :+ res.length()
+
+    if (keywords.contains(res)) ((KeywordToken(res), createLocation(start, end)), end.column)
+    else ((IdentifierToken(res), createLocation(start, end)), end.column)
+  }
+}
+
+object Lexer {
+  def apply(code: String, options: LexerOptions) =
+    new Lexer(code, options)
+
+  private val keywords = List[String](
+    "def", "defn", "let", "if", "match" // TODO: Add more if need
+  )
+
+  private val punctuations = List[Char](
+    '(', ')', ' ', '\t', '\n', '\r', ':', '[', ']', '{', '}' // TODO: Are these symbols enough?
```

Comment:
do +-*/ these mathematical symbols belongs to Lexer‘s duty？

Reply:
I treat symbols `+-*/` (and so on) like identifiers. Then we can allow users to define their own operators.